packages/xchain-doge/src/utils.ts

```diff
@@ -0,0 +1,295 @@
+    }
+  }
+}
+
+/**
+ * Get the balances of an address.
+ *
+ * @param {AddressParams} params
+ * @returns {Balance[]} The balances of the given address.
+ */
+export const getBalance = async (params: AddressParams): Promise<Balance[]> => {
+  try {
+    const balance = await sochain.getBalance(params)
+    return [
+      {
+        asset: AssetDOGE,
+        amount: balance,
+      },
+    ]
+  } catch (error) {
+    throw new Error(`Could not get balances for address ${params.address}`)
+  }
+}
+
+/**
+ * Validate the Doge address.
+ *
+ * @param {string} address
+ * @param {Network} network
+ * @returns {boolean} `true` or `false`.
+ */
+export const validateAddress = (address: Address, network: Network): boolean => {
+  try {
+    Dogecoin.address.toOutputScript(address, dogeNetwork(network))
+    return true
+  } catch (error) {
+    return false
+  }
+}
+
+/**
+ * Scan UTXOs from sochain.
+ *
+ * @param {AddressParams} params
+ * @returns {UTXO[]} The UTXOs of the given address.
+ */
+export const scanUTXOs = async (params: AddressParams): Promise<UTXO[]> => {
+  const unspent: DogeAddressUTXO[] = await sochain.getUnspentTxs(params)
+  const utxos: UTXO[] = []
+  for (const utxo of unspent) {
+    utxos.push({
+      hash: utxo.txid,
+      index: utxo.output_no,
+      value: assetToBase(assetAmount(utxo.value, DOGE_DECIMAL)).amount().toNumber(),
+    })
+  }
+
+  return utxos
+}
+
+/**
+ * Build transcation.
+ *
+ * @param {BuildParams} params The transaction build options.
+ * @returns {Transaction}
+ */
+export const buildTx = async ({
+  amount,
+  recipient,
+  memo,
+  feeRate,
+  sender,
+  network,
+  sochainUrl,
+}: TxParams & {
+  feeRate: FeeRate
+  sender: Address
+  network: Network
+  sochainUrl: string
+}): Promise<{ psbt: Dogecoin.Psbt; utxos: UTXO[] }> => {
+  if (!validateAddress(recipient, network)) throw new Error('Invalid address')
+
+  const utxos = await scanUTXOs({ sochainUrl, network, address: sender })
+  if (utxos.length === 0) throw new Error('No utxos to send')
+
+  const feeRateWhole = Number(feeRate.toFixed(0))
+  const compiledMemo = memo ? compileMemo(memo) : null
+
+  const targetOutputs = []
+  //1. output to recipient
+  targetOutputs.push({
+    address: recipient,
+    value: amount.amount().toNumber(),
+  })
+  //2. add output memo to targets (optional)
+  if (compiledMemo) {
+    targetOutputs.push({ script: compiledMemo, value: 0 })
+  }
+  const { inputs, outputs } = accumulative(utxos, targetOutputs, feeRateWhole)
+
+  // .inputs and .outputs will be undefined if no solution was found
+  if (!inputs || !outputs) throw new Error('Balance insufficient for transaction')
+
+  const psbt = new Dogecoin.Psbt({ network: dogeNetwork(network) }) // Network-specific
+  // TODO: Doge recommended fees is greater than the recommended by Bitcoinjs-lib (for BTC),
+  //       so we need to increase the maximum fee rate. Currently, the fast rate fee is near ~650000sats/byte
+  psbt.setMaximumFeeRate(650000)
```

Comment:
1. You are right, actually is not a TODO. I preferred to set a default fee value instead of calculating in run time. There could be a better approach for this.
2. I set it as the maximum fee rate. Usually, the fast fee rate average doesn't be greater than 600000 sats/byte
3. Yes, it is. By default, xchain-doge will use the lowest fee rate (100000 - 200000 sats/byte)

Reply:
Ok.